### ISSUE TITLE
Multiline documentation paragraphs get bad linebreaks in documentation hover

### ISSUE THREAD

**Issue by oblitum** [neoclide/coc.nvim#3720]
## Describe the bug

The linebreaks from documentation are transported to the hover docs, the hover window can often be narrower than the line length in documentation, so this creates spurious linebreaks in addition to the ones that exists due to window size.

## Reproduce the bug

**We will close your issue when you don't provide minimal vimrc and we can't
reproduce it**

- Create file `mini.vim` with：

  ```vim
  set nocompatible
  set runtimepath^=/path/to/coc.nvim
  filetype plugin indent on
  syntax on
  set hidden
  ```

- Install `coc-rust-analyzer`

- Start (neo)vim with command: `vim -u mini.vim`

- File edited in screenshot bellow is https://docs.rs/bevy_transform/0.6.0/src/bevy_transform/components/transform.rs.html#111-118.

- Check documentation hover (better reproduced if it's small)

## Screenshots (optional)

![img-2022-03-21-210527](https://user-images.githubusercontent.com/1269815/159382847-ab4efcf3-ef60-4577-a311-5f4b1136114e.png)

**Comment by yaegassy** [comment]
This is the result of `textDocument/hover`, but since the line is broken where the `\n` is in, it seems to be normal behavior.

```
[Trace - 12:57:50 PM] Received response 'textDocument/hover - (19)' in 5ms.
Result: {
    "contents": {
        "kind": "markdown",
        "value": "\n```rust\nbevy_transform::components::transform::Transform\n```\n\n```rust\npub fn looking_at(self, target: Vec3, up: Vec3) -> Self\n```\n\n---\n\nUpdates and returns this [`Transform`](https://docs.rs/bevy_transform/0.6.0/bevy_transform/components/transform/struct.Transform.html) by rotating it so that its unit vector in the\nlocal z direction is toward `target` and its unit vector in the local y direction\nis toward `up`."
    },
    "range": {
        "start": {
            "line": 3,
            "character": 37
        },
        "end": {
            "line": 3,
            "character": 47
        }
    }
}
```

**Comment by chemzqm** [comment]
Configure `maxWidth` in your `"hover.floatConfig"` configuration.

**Comment by oblitum** [comment]
> This is the result of `textDocument/hover`, but it appears to be normal behavior since it is carriage-returned at the `\n` location.

Indeed, to be honest, I think this is a duplicate, I think I've gone through this before and was against applying "formatting" to `Result.contents.value`.

> Configure `maxWidth` in your `"hover.floatConfig"` configuration.

It's an option, but I wished better flowed text in smaller-by-default floats. Large floats get much in the way coding.

**Comment by chemzqm** [comment]
The language server should be responsible to format the doucment.

**Comment by oblitum** [comment]
Right, agree, it should be a fix in rust-analyzer.

**Comment by oblitum** [comment]
bump?

**Comment by chemzqm** [comment]
Not supported by marked, you can try to hack marked if you want the lines to get merged, can't be done with custome `render` of `marked`.

**Comment by oblitum** [comment]
Ok, thanks. Well, then maybe they need this fixed upstream, I'll check out what's up.

**Comment by chemzqm** [comment]
```js
const {marked} = require('marked')

marked.setOptions({
  gfm: true,
  breaks: true
})

const html = marked.parse('foo\nbar')
```
results:
```
<p>foo<br>bar</p>
```
Same behavior on github.

**Comment by oblitum** [comment]
> Same behavior on github.

Do you mean github comments? Indeed that's the case, but I suspect they're not following original/standard markdown guidelines on purpose because people in general wish a newline when they insert one when commenting, instead of having to learn that they have to end line with 2 spaces or more for hard line break.

See this issue for example:

- https://github.com/markedjs/marked/issues/1380

Possibly coc.nvim should adopt `breaks: false` to have standard markdown soft and hard line breaks? I think this is the best behavior for Markdown in documentation context.

**Comment by chemzqm** [comment]
That option possible cause unnecessary new lines as well.

**Comment by oblitum** [comment]
Yeah, I'm gonna have to check it on my dev machine asap. Their docs seem confusing indeed, because it says it should be `false` by default (supposedly normal markdown behavior, requiring two spaces), which was the case when I opened this issue. It's now explicitly `true` after late commits, which would generate line breaks on new lines alone, according to the linked issue. In any case, `marked` has to support original and CommonMark markdown somehow at this, because they support that, have include tests for this, otherwise they would be support GFM as in GitHub comments solely.

**Comment by oblitum** [comment]
Ok, just tried marked locally and it doesn't generate `<br>` when `breaks: false`, which is the default (I also checked it when by simply not setting the option). Only when `breaks: true` it generates `<br>`. What happens is that a `\n` is left as-is when `breaks: false`, and when `breaks: true`, then the `\n` is consumed in the output, and transformed into a `<br>`. If when consider how HTML works, a `<br>` presents as a new line, but a `\n` will simply not show up, it only exists in the HTML source. So this is the confusion here, not sure how it should be handled, but I think it would be ideal to have rendering of newlines only in case of `<br>` output it seems...

**Comment by chemzqm** [comment]
`breaks: true` works better with unnecessary lines get removed, you can't simply replace `\n` with empty space to fix the issue.
IMO, the line break makes the text easier to read on most cased, configure `"maxWidth"` for you window should help  to avoid your issue.

**Comment by oblitum** [comment]
> `breaks: true` works better with unnecessary lines get removed

Okay, I'm not sure which ones are removed though. At the moment, either ending a line with two spaces or not (which results either in `<br>` or `\n`, respectively) will generate line breaks in documentation float, so I don't see any case where they get removed, in any case it simply results in line breaks.

> IMO, the line break makes the text easier to read on most cases, configure "maxWidth" for you window should help to avoid your issue.

I realize what you mean, it's some kind of tension between honoring the usual Markdown specs (GFM-_comments_ is unusual) vs preserving the original text layout as in the source code comments, even though by that you can break _usual_ Markdown rendering rules. If there was a way to render it properly, I'd prefer to have Markdown honored, but even then, it seems you would prefer to preserve line breaks from comment sources, for the to display verbatim in a large window.

**Comment by cheukyin699** [comment]
Hello, just joining in. I was wondering if we could make the `breaks: true` or `breaks: false` a preference, which by default goes to `breaks: true`.

Something like this (untested)? https://github.com/cheukyin699/coc.nvim/commit/1e1c057696049fd694ab197c69ff1803e243bb78

**Comment by chemzqm** [comment]
Add configuration `coc.preferences.enableGFMBreaksInMarkdownDocument`
